### ISSUE TITLE
rank 0 GPU OOM while  multi-machine training with DDP

### ISSUE THREAD

**Issue by ospeech** [wenet-e2e/wenet#989]
not sure this issue is usual.
when multi-machine training with two 8-gpus(Nvidia 3090), the rank 0 (the first one in master machine) process will be killed by system after several hours. I have monitored the memory using of each gpu. the rank 0 require a lot of system memory approach the break point. refer as follows:
2022-03-23 23:57:28,855 DEBUG TRAIN Batch 0/185100 loss 4.577072 loss_att 4.222489 loss_ctc 5.404430 lr 0.00065741 rank 0 RSS MEM USE: 5.11G
2022-03-23 23:57:43,632 DEBUG TRAIN Batch 0/185200 loss 5.362794 loss_att 4.925412 loss_ctc 6.383353 lr 0.00065721 rank 0 RSS MEM USE: 5.11G
2022-03-23 23:57:57,979 DEBUG TRAIN Batch 0/185300 loss 7.436929 loss_att 6.969763 loss_ctc 8.526984 lr 0.00065704 rank 0 RSS MEM USE: 5.11G
2022-03-23 23:58:12,914 DEBUG TRAIN Batch 0/185400 loss 9.745964 loss_att 7.788527 loss_ctc 14.313314 lr 0.00065687 rank 0 RSS MEM USE: 46.51G
2022-03-23 23:58:27,353 DEBUG TRAIN Batch 0/185500 loss 7.055099 loss_att 5.558590 loss_ctc 10.546951 lr 0.00065670 rank 0 RSS MEM USE: 105.61G
2022-03-23 23:58:41,824 DEBUG TRAIN Batch 0/185600 loss 21.082787 loss_att 17.530172 loss_ctc 29.372223 lr 0.00065650 rank 0 RSS MEM USE: 161.09G
2022-03-23 23:58:56,216 DEBUG TRAIN Batch 0/185700 loss 4.324765 loss_att 3.620107 loss_ctc 5.968966 lr 0.00065633 rank 0 RSS MEM USE: 212.04G
2022-03-23 23:59:11,208 DEBUG TRAIN Batch 0/185800 loss 5.875733 loss_att 5.488202 loss_ctc 6.779974 lr 0.00065616 rank 0 RSS MEM USE: 268.65G
2022-03-23 23:59:25,299 DEBUG TRAIN Batch 0/185900 loss 7.695665 loss_att 6.738754 loss_ctc 9.928459 lr 0.00065600 rank 0 RSS MEM USE: 314.05G
2022-03-23 23:59:40,244 DEBUG TRAIN Batch 0/186000 loss 9.930955 loss_att 8.650377 loss_ctc 12.918968 lr 0.00065580 rank 0 RSS MEM USE: 368.19G
2022-03-23 23:59:53,885 DEBUG TRAIN Batch 0/186100 loss 9.757288 loss_att 8.881466 loss_ctc 11.800873 lr 0.00065563 rank 0 RSS MEM USE: **412.03G**
2022-03-24 00:00:07,981 DEBUG TRAIN Batch 0/186200 loss 10.991848 loss_att 8.944845 loss_ctc 15.768186 lr 0.00065546 rank 0 RSS MEM USE: **462.48G**
2022-03-24 00:00:21,908 DEBUG TRAIN Batch 0/186300 loss 5.258793 loss_att 4.664874 loss_ctc 6.644607 lr 0.00065529 rank 0 RSS MEM USE: **508.47G**
2022-03-24 00:00:37,008 DEBUG TRAIN Batch 0/186400 loss 6.789611 loss_att 6.216270 loss_ctc 8.127404 lr 0.00065509 rank 0 RSS MEM USE: 560.11G
2022-03-24 00:00:51,764 DEBUG TRAIN Batch 0/186500 loss 7.544255 loss_att 6.040744 loss_ctc 11.052447 lr 0.00065493 rank 0 RSS MEM USE: **616.43G**
2022-03-24 00:01:07,753 DEBUG TRAIN Batch 0/186600 loss 7.861217 loss_att 6.178774 loss_ctc 11.786919 lr 0.00065476 rank 0 RSS MEM USE: **672.18G**
2022-03-24 00:01:48,785 DEBUG TRAIN Batch 0/186700 loss 10.188281 loss_att 8.669827 loss_ctc 13.731339 lr 0.00065459 rank 0 RSS MEM USE: **707.11G**

-------------------------
some key infos:
dist_backend="nccl"
init_method="tcp://_ip:port_"
pytorch version: 1.10
------------------------
i have tried:
1. change the master machine
2. change the rank 0 GPU
3. reduce the worker_number

still same issue occur.  the batch number of break point is random and single machine train with same traing data is ok.

**Comment by robin1001** [comment]
I think you can try:
1. make it works on single card
2. reduce the batch_size

**Comment by ospeech** [comment]
> I think you can try:
> 
> 1. make it works on single card
> 2. reduce the batch_size

i train with same trainging data in single node, which is normal.   it didnot work to reduce the batch size and worker number. 

Yesterday, i got following error after 14 epochs:
envs/wenet/lib/python3.8/multiprocessing/process.py:108: ResourceWarning: unclosed file <_io.BufferedReader name='xxxxxxx/shards_000000512.tar'>

**Comment by ospeech** [comment]
it runs normally with a wenet-docker cluster.maybe there is something wrong related to hardward, which is hard to find out. 

close the issue!
